### DISCUSSION TITLE
Has anyone successfully compiled into WebAssembly?

### DISCUSSION THREAD

**Discussion by broomva** [fastapi/fastapi#12228] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
We're talking of a regular API server with router

from fastapi import Depends, FastAPI, APIRouter

base_router = APIRouter()

app = FastAPI(
    title=PROJECT_NAME,
    debug=DEBUG,
    version=VERSION,
    lifespan=lifespan,
    dependencies=[Depends(get_api_key)],
)
app.include_router(router, prefix=API_PREFIX)

auth_scheme = HTTPBearer()


@app.get("/")
async def redirect_root_to_docs():
    return RedirectResponse("/docs")


# Set all CORS enabled origins
app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
    expose_headers=["*"],
)

base_router.include_router(faucet.router, tags=["faucet"], prefix="/v1")
... other routes
```

### Description

I want to deploy my endpoints in ICP's (https://internetcomputer.org/) decentralized hosting, but there are some challenges related to compiling the code into wasm to be used inside ICP's cannisters. I don't even know if it is possible, but it would be great to achive this and be able to host a server inside a decentralized architecture

Using a Python WebAssembly (Wasm) Runtime
	• How it works: ICP supports WebAssembly (Wasm), which allows code from multiple languages to run on the network. You can compile Python code to WebAssembly using tools like Pyodide or wasmer. However, FastAPI would require additional work since it relies on Python’s asyncio and web frameworks that aren’t directly portable to Wasm.
	• Benefit: This allows you to run Python code within ICP’s canisters, effectively decentralizing your backend API.
	• Challenge: FastAPI is not designed to run inside a Wasm environment out of the box, so you may need to strip down or rework parts of your API and possibly write an ICP-specific bridge layer.

### Operating System

Linux

### FastAPI Version

0.112.2

### Pydantic Version

1.10.18

### Python Version

Python 3.11.9

**Comment by syrusakbary** [comment]
> FastAPI would require additional work since it relies on Python’s asyncio and web frameworks that aren’t directly portable to Wasm

Hey, Syrus here from Wasmer.
Almost all Python APIs (including asyncio and threads) are usable within WebAssembly thanks to [WASIX](http://wasix.org).
We also launched the py2wasm utility to [convert Python to WebAssembly](https://wasmer.io/posts/py2wasm-a-python-to-wasm-compiler).
